test/SlowTests/Issues/RavenDB-15754.cs

```diff
@@ -0,0 +1,259 @@
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using FastTests;
+using Raven.Client.Documents;
+using Raven.Client.Documents.Changes;
+using Raven.Client.Documents.Indexes;
+using Raven.Server.Config;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_15754 : RavenTestBase
+    {
+        public RavenDB_15754(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        private const int _employeesCount = 20_000;
+        private const string _managedAllocationsBatchLimit = "16";
+
+        [Fact]
+        public async Task CanIndexReferencedDocumentChange()
+        {
+            using (var store = GetDocumentStore(new Options
+            {
+                ModifyDatabaseRecord = x => x.Settings[RavenConfiguration.GetKey(x => x.Indexing.ManagedAllocationsBatchLimit)] = _managedAllocationsBatchLimit
+            }))
+            {
+                const string companyName1 = "Hibernating Rhinos";
+                const string companyName2 = "HR";
+                var company = new Company
+                {
+                    Name = companyName1
+                };
+
+                using (var session = store.OpenAsyncSession())
+                {
+                    await session.StoreAsync(company);
+                    await session.SaveChangesAsync();
+
+                    using (var bulk = store.BulkInsert())
+                    {
+                        for (var i = 0; i < _employeesCount; i++)
+                        {
+                            await bulk.StoreAsync(new Employee
+                            {
+                                CompanyId = company.Id
+                            });
+                        }
+                    }
+                }
+
+                var index = new Index();
+                new Index().Execute(store);
+
+                WaitForIndexing(store, timeout: TimeSpan.FromMinutes(3));
+                await AssertCount(store, companyName1, _employeesCount);
+
+                var batchCount = 0;
```

Comment:
This is racy, even if we did have 5 batches, we may not get notified on them in time, no?

Reply:
fixed